### PR TITLE
Beta Release 28 - Includes up to Developer Release 5723

### PR DESCRIPTION
**Pull request by cozza13**
http://rawgit.com/cozza13/QAtesting/master/QAofRC.js

* Use matching hash keys for input/output when batch-loading scripts.
* Fix some crash on exits
* Don't send audio/avatar data for ignoree to ignored node
* Introduce Instanced Stereo Drawcall
  * Introduces the instanced stereo drawcall technique to acheive the drawcall of a shape in stereo to the left and right veiwports of the stereo framebuffer in a single instanced drawcall
* Disable threaded reprojection when running oculus in OpenVR
* Fix for mac min openGL version
* Block scripts until follow-on includes are finished but still avoid multiple evaluation of included urls
* Guard patched agents asking for other agents in domain-server
* Improve Input Stage Format switching in GL45Backend
* Fix - When avatar's parent is deleted, avatar does not move
* Fix buffer leak in line entity
* Provide tab-like control over whether suggestions shows snapshots, place concurrency, or all. 
* Fix polyvox memory leak
* Bug fix for blank webEntities during long sessions
* Multichannel audio support
* Animated GIF Snapshots
* Turn Edit.js on when importing entities
* Fix polyvox buffer size
* Add hardware address based permissions to domain-server

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: e9563ef9af82015ebeffcd8718dac7d747fcbabd available at: https://gist.github.com/2b7a92141739cbca9f8c9ab4f7024d8c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9098-e9563ef.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9098-e9563ef.dmg

File: libraries/entities-renderer/src/RenderableWebEntityItem.cpp [hunk 1]
```diff
@@ -31,7 +31,7 @@
 const float METERS_TO_INCHES = 39.3701f;
 static uint32_t _currentWebCount { 0 };
 // Don't allow more than 100 concurrent web views
```
Comment (thoys) [inline libraries/entities-renderer/src/RenderableWebEntityItem.cpp]: ???

**Comment by Nex-Pro** [comment]
Testing this.

**Comment by kevinmthomasse** [comment]
testing too

**Comment by kevinmthomasse** [comment]
## QA Test
**QA Tester**: KevinMThomas
**localhost 'createTestBlocks.js' Result**: **localhost 'createTestBlocks.js'** SUCCESS
**localhost ATP Upload + Edit Result**: **localhost ATP** SUCCESS
**Echo Server Audio**: **Echo Server Audio** SUCCESS
**localhost 'harmonicOscillator.js' Result**: **localhost 'harmonicOscillator.js'** UNTESTED
**localhost 'reverbTest.js' Result**: **localhost 'reverbTest.js'** SUCCESS
**localhost 'playaPerformanceTest.js' Result**: **localhost 'playaPerformanceTest.js'** SUCCESS
**dev-playa 'playaPerformanceTest.js' Result**: **dev-playa 'playaPerformanceTest.js'** SUCCESS
**dev-demo 'playaPerformanceTest.js' Result**: **dev-demo 'playaPerformanceTest.js'** SUCCESS
**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**: Successful test with Hydras.

**Domain** |localhost |dev-playa |dev-demo
---|---|---|---
**Render Rate**: |60 |60 |60
**Present Rate**: |60 |60 |60
**Present New Rate**: |52 |54 |58
**CPU Buffer Count**: |2189 |1211 |3928
**GPU Buffer Count**: |756 |354 |797
**CPU Memory Usage**: |28631040 |21696512 |69292032
**GPU Memory Usage**: |19296256 |12980224 |31584256
**CPU Texture Count**: |221 |85 |344
**GPU Texture Count**: |196 |63 |234
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |41183600 |40045008 |367258064
**GPU Texture Memory Usage**: |588473537 |221962070 |1003008100
**GPU Virtual Memory Usage**: |575404101|220503858|986006432
**Triangle Count**: |259295 |154037 |439875
**Triangle Rate**: |16205937 |9627312 |27492187
**API Drawcall Count**: |198 |124 |227
**Drawcall Count**: |2248 |120 |1150
**Drawcall Rate**: |140500 |7500 |71875
**State Changes Texture Count**:|359 |149 |308
**Pipeline Count**: |65 |44 |50
**Item Opaques**: |268 |222 |240
**Item Translucents**: |88 |80 |46
**Item Lights**: |16 |76 |16
**Timing Opaques**: |1.089 |0.479 |0.787
**Timing Translucents**: |0.14 |0.198 |0.16
**Timing Lighting**: |0.085 |0.104 |0.066
**Render Frame**: |2.219 |1.822 |1.881

**RC Specific tests**

**Test** | **Description** | **Results**
---|---|---
**Test 1**|For performance testing use. Localhost - dev-playa - dev-demo|Success
**Test 2**|Install a fresh copy of "Home' and do tutorial|Success
**Test 3**|Go to localhost - Make sure you can hear the background sounds.|Success
**Test 4**|Launch firework, on dev-playa, can other hear?|Untested
**Test 5**|Switch avatar and make sure it does not crash|Success
**Test 6**|Test group functions|Untested
**Test 7**|Test Kick now includes MAC address|Untested
**Test 8**|Enrure that you can distance grab|Success
**Test 9**|Bring up and edit an objects with hand controllers|Partial success with Hydras as was able to work the menus however unable to X out a UI window with the Hydra ongoing issue not related to PR.

**Comment by OmegaHeron** [comment]
Windows 10 - i5 6500 CPU - GTX970 GPU

## QA Test
**QA Tester**: OmegaHeron
**localhost 'createTestBlocks.js' Result**: **localhost 'createTestBlocks.js'** SUCCESS
**localhost ATP Upload + Edit Result**: **localhost ATP** SUCCESS
**Echo Server Audio**: **Echo Server Audio** SUCCESS
**localhost 'harmonicOscillator.js' Result**: **localhost 'harmonicOscillator.js'** SUCCESS
**localhost 'reverbTest.js' Result**: **localhost 'reverbTest.js'** SUCCESS
**localhost 'playaPerformanceTest.js' Result**: **localhost 'playaPerformanceTest.js'** SUCCESS
**dev-playa 'playaPerformanceTest.js' Result**: **dev-playa 'playaPerformanceTest.js'** SUCCESS
**dev-demo 'playaPerformanceTest.js' Result**: **dev-demo 'playaPerformanceTest.js'** SUCCESS
**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**:

**RC Specific tests**

**Test** | **Description** | **Results**
---|---|---
**Test 1**|For performance testing use. Localhost - dev-playa - dev-demo|Yes
**Test 2**|Install a fresh copy of "Home' and do tutorial|Pass
**Test 3**|Go to localhost - Make sure you can hear the background sounds.|Pass
**Test 4**|Launch firework, on dev-playa, can other hear?|N/A
**Test 5**|Switch avatar and make sure it does not crash|Pass
**Test 6**|Test group functions|Pass
**Test 7**|Test Kick now includes MAC address|Pass
**Test 8**|Enrure that you can distance grab|N/A
**Test 9**|Bring up and edit an objects with hand controllers|N/A

Desktop Mode - 1600x900 Maximized Window (vsync lock on)

**Domain** |localhost |dev-playa |dev-demo
---|---|---|---
**Render Rate**: |60 |60 |60
**Present Rate**: |60 |60 |60
**Present New Rate**: |60 |60 |60
**CPU Buffer Count**: |2320 |4157 |3646
**GPU Buffer Count**: |823 |1405 |966
**CPU Memory Usage**: |37707776 |73289728 |64016384
**GPU Memory Usage**: |27467776 |48168960 |30183424
**CPU Texture Count**: |223 |330 |301
**GPU Texture Count**: |190 |275 |264
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |75099312 |246549776 |117767440
**GPU Texture Memory Usage**: |577306197 |1125566703 |1081477628
**GPU Virtual Memory Usage**: |563586826|1230287331|1062351497
**Triangle Count**: |247570 |851503 |342610
**Triangle Rate**: |16504666 |50088411 |21413125
**API Drawcall Count**: |190 |329 |204
**Drawcall Count**: |1453 |372 |1125
**Drawcall Rate**: |96866 |21882 |70312
**State Changes Texture Count**:|361 |337 |286
**Pipeline Count**: |69 |40 |52
**Item Opaques**: |254 |468 |218
**Item Translucents**: |92 |59 |44
**Item Lights**: |16 |82 |14
**Timing Opaques**: |1.32 |1.364 |0.84
**Timing Translucents**: |0.178 |0.072 |0.159
**Timing Lighting**: |0.16 |0.11 |0.288
**Render Frame**: |2.553 |2.665 |2.35

Desktop Mode - 1600x900 Maximized Window (vsync unlocked)

**Domain** |localhost |dev-playa |dev-demo
---|---|---|---
**Render Rate**: |318 |209 |338
**Present Rate**: |318 |209 |338
**Present New Rate**: |233 |180 |267
**CPU Buffer Count**: |2722 |4494 |3513
**GPU Buffer Count**: |923 |1440 |922
**CPU Memory Usage**: |32030720 |82026496 |63483904
**GPU Memory Usage**: |18468864 |57606144 |30306304
**CPU Texture Count**: |229 |296 |300
**GPU Texture Count**: |194 |255 |259
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |80685424 |166332880 |126679504
**GPU Texture Memory Usage**: |576040453 |1134766234 |1068415400
**GPU Virtual Memory Usage**: |562452167|1123439046|1050119433
**Triangle Count**: |250084 |836328 |354292
**Triangle Rate**: |62521000 |209082000 |118097333
**API Drawcall Count**: |196 |327 |209
**Drawcall Count**: |3007 |370 |1115
**Drawcall Rate**: |751750 |92500 |371666
**State Changes Texture Count**:|363 |338 |290
**Pipeline Count**: |69 |37 |50
**Item Opaques**: |259 |465 |221
**Item Translucents**: |93 |60 |44
**Item Lights**: |16 |80 |14
**Timing Opaques**: |1.128 |1.569 |0.89
**Timing Translucents**: |0.227 |0.092 |0.15
**Timing Lighting**: |0.129 |0.169 |0.085
**Render Frame**: |2.531 |3.059 |1.859

DK2

**Domain** |localhost |dev-playa |dev-demo
---|---|---|---
**Render Rate**: |75 |75 |75
**Present Rate**: |75 |75 |75
**Present New Rate**: |75 |75 |75
**CPU Buffer Count**: |2413 |3797 |4132
**GPU Buffer Count**: |731 |787 |852
**CPU Memory Usage**: |27283456 |64253952 |84934656
**GPU Memory Usage**: |12632064 |18796544 |25653248
**CPU Texture Count**: |233 |292 |352
**GPU Texture Count**: |197 |165 |257
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |82521264 |426347792 |324791568
**GPU Texture Memory Usage**: |757000085 |942924330 |990589757
**GPU Virtual Memory Usage**: |743095024|934939545|977777135
**Triangle Count**: |441908 |1770414 |783035
**Triangle Rate**: |31564857 |136185692 |65252916
**API Drawcall Count**: |162 |313 |222
**Drawcall Count**: |2267 |699 |2267
**Drawcall Rate**: |161928 |53769 |188916
**State Changes Texture Count**:|305 |310 |301
**Pipeline Count**: |62 |37 |52
**Item Opaques**: |213 |451 |240
**Item Translucents**: |89 |36 |47
**Item Lights**: |16 |85 |14
**Timing Opaques**: |0.885 |1.671 |1.069
**Timing Translucents**: |0.157 |0.06 |0.139
**Timing Lighting**: |0.171 |0.24 |0.21
**Render Frame**: |2.403 |3.628 |2.632

CPU comparison to last RC (5695) -- Red this PR, Blue last RC.

5695vs9098r5

Tested multiple web entities -- Placed 21 and 20 showed (as expected).

Checked AC to DS connection perms work as expected.

Checked for memory leaks - none detected.

Icon for kick/mute still remains after kicking an AV -- it "floats" away with de-rezzing AV, but remains in scene.  Image one -- view from where AV was kicked - not red kick/mute overlay in distance -- Image two close up of ghosted kick/mute overlay.  Persists until Interface is restarted.

hifi-snap-by-omegaheron-on-2016-11-19_12-59-48

hifi-snap-by-omegaheron-on-2016-11-19_13-00-06

**Comment by Nex-Pro** [comment]
### PR #9098 Beta Release 28

**Testing Architecture:** Desktop PC
**CPU:** 4.4 ghz Intel i7
**RAM:** 16 GB
**GPU:** Nvidia gtx 1070 8gb

**Testing OS:** win 10 pro

**HMD:** HTC Vive.
**Hand Controllers:** Vive


## QA Test

**QA Tester**: Piper.Peppercorn
**localhost 'createTestBlocks.js' Result**: **SUCCESS**
**localhost ATP Upload + Edit Result**: **SUCCESS**
**Echo Server Audio**: **SUCCESS**
**localhost 'harmonicOscillator.js' Result**: **SUCCESS**
**localhost 'reverbTest.js' Result**: **SUCCESS**
**localhost 'playaPerformanceTest.js' Result**: **SUCCESS**
**dev-playa 'playaPerformanceTest.js' Result**: **SUCCESS**
**dev-demo 'playaPerformanceTest.js' Result**: **SUCCESS**
**QA Pass**:**SUCCESS**

**Additional Test Results + Observations**:

Still having issue with render rate going uncapped after switching to desktop.

Spent some time testing the voxel paint system. Seems fine.


**Domain** |localhost |dev-playa |dev-demo
---|---|---|---
**Render Rate**: |90 |90 |90
**Present Rate**: |90 |89 |90
**Present New Rate**: |64 |79 |90
**CPU Buffer Count**: |3650 |4565 |4579
**GPU Buffer Count**: |1702 |1209 |1089
**CPU Memory Usage**: |56020992 |96468992 |89083904
**GPU Memory Usage**: |36716544 |49758208 |42659840
**CPU Texture Count**: |286 |300 |359
**GPU Texture Count**: |239 |225 |305
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |105003088 |329846032 |227077392
**GPU Texture Memory Usage**: |1171243908 |1260268082 |1352481605
**GPU Virtual Memory Usage**: |1344667764|1391995045|1576696095
**Triangle Count**: |245940 |2040222 |1178136
**Triangle Rate**: |22358181 |185474727 |107103272
**API Drawcall Count**: |43 |324 |256
**Drawcall Count**: |5085 |721 |2315
**Drawcall Rate**: |462272 |65545 |210454
**State Changes Texture Count**:|51 |326 |354
**Pipeline Count**: |40 |38 |57
**Item Opaques**: |30 |470 |279
**Item Translucents**: |17 |43 |53
**Item Lights**: |0 |85 |14
**Timing Opaques**: |0.256 |1.547 |1.131
**Timing Translucents**: |0.027 |0.048 |0.167
**Timing Lighting**: |0.146 |0.236 |0.245
**Render Frame**: |1.413 |3.336 |2.819


**RC Specific tests**

**Test** | **Description** | **Results**
**Test 1**|For performance testing use. Localhost - dev-playa - dev-demo|**CHECK**
**Test 2**|Install a fresh copy of "Home' and do tutorial|**SUCCESS**
**Test 3**|Go to localhost - Make sure you can hear the background sounds.|**SUCCESS**
**Test 4**|Launch firework, on dev-playa, can other hear?|**SUCCESS**
**Test 5**|Switch avatar and make sure it does not crash|**SUCCESS**
**Test 6**|Test group functions|**UNTESTED**
**Test 7**|Test Kick now includes MAC address|**UNTESTED**
**Test 8**|Enrure that you can distance grab|**SUCCESS**
**Test 9**|Bring up and edit an objects with hand controllers|**SUCCESS**
**Test 10**|Grab objects in dev-demo|**SUCCESS**

**Comment by hifi-gustavo** [comment]
Build console output for commit: 14857a085a4fb15ae43c3c54a76c10be64bb97d3 available at: https://gist.github.com/2b7a92141739cbca9f8c9ab4f7024d8c<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9098-14857a0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9098-14857a0.dmg

**Comment by OmegaHeron** [comment]
Retest for Personal Space functions - Windows 10 - i5 6500 CPU - GTX 970 GPU

Repeated basic tests, benchmarks and specific personal space tests from #9099 : **Pass**

**Comment by OmegaHeron** [comment]
I had made comment in my initial QA about kick/mute icons remaining "ghosted".

Incorrect - icon is **not** left ghosted.  Here's actually what happened.  In localhost content when I engaged kick I never noticed there was a distant second kick set for scripted agent in domain.. I kicked my test second agent (near me) *then* noticed a remaining kick icon that's there for the scripted agent at 0,0,0 due to *kick button still being engaged*.  No bug there.

**Pull request merged**